File: crawler-files/crawl.js
#!/usr/bin/env node
/**
 * Interactive Website Crawler CLI
 * Easy-to-use wrapper for crawling any website
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');
const axios = require('axios');
const cheerio = require('cheerio');
const { URL } = require('url');
const pLimit = require('p-limit');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function question(prompt) {
  return new Promise(resolve => {
    rl.question(prompt, resolve);
  });
}

class AdvancedWebsiteCrawler {
  constructor(startUrl, options = {}) {
    this.startUrl = startUrl.replace(/\/$/, '');
    this.outputDir = options.outputDir || 'website_crawl';
    this.followExternal = options.followExternal !== false;
    this.maxPages = options.maxPages || null;
    this.maxConcurrent = options.maxConcurrent || 5;
    this.timeout = options.timeout || 15000;
    this.verbose = options.verbose !== false;
    
    // Parse domain
    const parsed = new URL(startUrl);
    this.domain = `${parsed.protocol}//${parsed.hostname}`;
    this.hostname = parsed.hostname;
    this.protocol = parsed.protocol;
    
    // Tracking
    this.visited = new Map(); // url -> success/fail
    this.queue = new Set([startUrl]);
    this.domainsFound = new Set([this.hostname]);
    this.fileTypes = {};
    this.errors = [];
    this.stats = {
      pagesDownloaded: 0,
      filesDownloaded: 0,
      bytesSaved: 0,
      domainsVisited: 1,
      externalLinksFound: 0,
      errors: 0,
      startTime: Date.now(),
      skipped: 0
    };
    
    // Create output directory
    this.ensureDir(this.outputDir);
    
    // Axios instance
    this.client = axios.create({
      timeout: this.timeout,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
      },
      maxRedirects: 10,
      validateStatus: () => true // Accept all status codes
    });
  }

  log(msg) {
    if (this.verbose) {
      console.log(msg);
    }
  }

  ensureDir(dirPath) {
    if (!fs.existsSync(dirPath)) {
      fs.mkdirSync(dirPath, { recursive: true });
    }
  }

  isValidUrl(urlString) {
    try {
      const url = new URL(urlString);
      
      // Skip special URLs
      if (!url.protocol.startsWith('http')) {
        return false;
      }
      
      // Skip data URIs and common junk
      if (urlString.startsWith('data:') || urlString.startsWith('blob:')) {
        return false;
      }
      
      return true;
    } catch {
      return false;
    }
  }

  shouldFollowUrl(urlString) {
    try {
      const url = new URL(urlString);
      const hostname = url.hostname;
      
      if (this.followExternal) {
        return true; // Follow everything
      }
      
      // Stay on domain
      return hostname === this.hostname;
    } catch {
      return false;
    }
  }

  getLocalPath(urlString) {
    try {
      const url = new URL(urlString);
      let pathname = url.pathname.replace(/^\//, '');
      
      // Handle root
      if (!pathname || pathname.endsWith('/')) {
        pathname = path.join(pathname, 'index.html');
      }
      
      // Add extension if missing
      if (!path.extname(pathname)) {
        pathname += '/index.html';
      }
      
      // Handle query strings - append them to filename
      if (url.search) {
        const base = path.dirname(pathname);
        const ext = path.extname(pathname);
        const name = path.basename(pathname, ext);
        const query = url.search.slice(1).replace(/[&=]/g, '_').substring(0, 30);
        pathname = path.join(base, `${name}_${query}${ext}`);
      }
      
      return path.join(this.outputDir, pathname);
    } catch {
      return path.join(this.outputDir, 'error.html');
    }
  }

  getFileType(urlString) {
    try {
      const url = new URL(urlString);
      const ext = path.extname(url.pathname).toLowerCase();
      return ext || 'unknown';
    } catch {
      return 'unknown';
    }
  }

  isAssetUrl(urlString) {
    const assetExtensions = [
      '.css', '.js', '.png', '.jpg', '.jpeg', '.gif', '.svg', '.webp',
      '.woff', '.woff2', '.ttf', '.otf', '.eot',
      '.pdf', '.zip', '.mp4', '.mp3', '.webm', '.ico', '.xml', '.json',
      '.tar', '.gz', '.exe', '.dmg', '.iso'
    ];
    
    try {
      const url = new URL(urlString);
      const ext = path.extname(url.pathname).toLowerCase();
      return assetExtensions.includes(ext);
    } catch {
      return false;
    }
  }

  async downloadFile(urlString) {
    if (this.visited.has(urlString)) {
      return this.visited.get(urlString);
    }
    
    this.visited.set(urlString, false);
    
    try {
      const response = await this.client.get(urlString, {
        responseType: 'arraybuffer'
      });
      
      if (response.status < 200 || response.status >= 400) {
        throw new Error(`HTTP ${response.status}`);
      }
      
      const localPath = this.getLocalPath(urlString);
      this.ensureDir(path.dirname(localPath));
      
      fs.writeFileSync(localPath, response.data);
      
      const fileType = this.getFileType(urlString);
      this.fileTypes[fileType] = (this.fileTypes[fileType] || 0) + 1;
      
      this.stats.filesDownloaded++;
      this.stats.bytesSaved += response.data.length;
      
      this.log(`  ✓ ${urlString}`);
      this.visited.set(urlString, true);
      
      return { success: true, response, data: response.data };
      
    } catch (error) {
      const msg = `✗ ${urlString} (${error.message})`;
      this.log(`  ${msg}`);
      this.errors.push(msg);
      this.stats.errors++;
      this.visited.set(urlString, false);
      return { success: false };
    }
  }

  extractUrlsFromHtml(html, baseUrl) {
    const urls = new Set();
    
    try {
      const $ = cheerio.load(html);
      
      // Extract all links
      $('a[href]').each((_, el) => {
        let href = $(el).attr('href')?.trim();
        if (href && !href.startsWith('#') && !href.startsWith('javascript:') && !href.startsWith('mailto:')) {
          try {
            const absolute = new URL(href, baseUrl).href.split('#')[0];
            if (this.isValidUrl(absolute)) {
              urls.add(absolute);
            }
          } catch {}
        }
      });
      
      // Extract script/link/img sources
      $('script[src], link[href], img[src], source[src], video[src], audio[src]').each((_, el) => {
        let src = $(el).attr('src') || $(el).attr('href');
        src = src?.trim();
        if (src) {
          try {
            const absolute = new URL(src, baseUrl).href;
            if (this.isValidUrl(absolute)) {
              urls.add(absolute);
            }
          } catch {}
        }
      });
      
      // Extract from srcset
      $('[srcset]').each((_, el) => {
        const srcset = $(el).attr('srcset');
        if (srcset) {
          srcset.split(',').forEach(item => {
            const src = item.trim().split(/\s+/)[0];
            try {
              const absolute = new URL(src, baseUrl).href;
              if (this.isValidUrl(absolute)) {
                urls.add(absolute);
              }
            } catch {}
          });
        }
      });
      
      // Extract from data attributes
      $('[data-src], [data-href], [data-url]').each((_, el) => {
        const src = $(el).attr('data-src') || $(el).attr('data-href') || $(el).attr('data-url');
        if (src) {
          try {
            const absolute = new URL(src, baseUrl).href;
            if (this.isValidUrl(absolute)) {
              urls.add(absolute);
            }
          } catch {}
        }
      });
      
    } catch (error) {
      this.log(`  ⚠ Error parsing HTML from ${baseUrl}`);
    }
    
    return urls;
  }

  async crawl() {
    console.log('\n' + '='.repeat(70));
    console.log('🚀 WEBSITE CRAWLER STARTED');
    console.log('='.repeat(70));
    console.log(`📍 Start URL:      ${this.startUrl}`);
    console.log(`📁 Output Dir:     ${path.resolve(this.outputDir)}`);
    console.log(`🌍 Mode:           ${this.followExternal ? 'Follow ALL external links' : 'Stay on domain only'}`);
    console.log(`⚡ Concurrency:    ${this.maxConcurrent} simultaneous downloads`);
    console.log(`⏱️  Timeout:        ${this.timeout / 1000}s per request`);
    console.log('='.repeat(70) + '\n');
    
    const limit = pLimit(this.maxConcurrent);
    const promises = [];
    let processedBatch = 0;
    
    while (this.queue.size > 0) {
      if (this.maxPages && this.stats.pagesDownloaded >= this.maxPages) {
        console.log(`\n⏸️  Reached max pages limit (${this.maxPages})`);
        break;
      }
      
      // Process batch
      const batch = Array.from(this.queue).slice(0, this.maxConcurrent);
      console.log(`\n📦 Batch ${++processedBatch} (Queue: ${this.queue.size}, Downloaded: ${this.stats.filesDownloaded})`);
      
      for (const url of batch) {
        this.queue.delete(url);
        
        const promise = limit(async () => {
          if (this.visited.has(url)) return;
          
          const result = await this.downloadFile(url);
          if (!result.success) return;
          
          // Check if HTML
          const contentType = result.response.headers['content-type'] || '';
          if (!contentType.includes('text/html')) {
            return;
          }
          
          this.stats.pagesDownloaded++;
          
          // Extract new URLs
          try {
            const html = result.data.toString('utf-8');
            const newUrls = this.extractUrlsFromHtml(html, url);
            
            for (const newUrl of newUrls) {
              if (!this.visited.has(newUrl)) {
                // Track domains
                try {
                  const urlObj = new URL(newUrl);
                  if (!this.domainsFound.has(urlObj.hostname)) {
                    this.domainsFound.add(urlObj.hostname);
                    this.stats.domainsVisited++;
                    if (urlObj.hostname !== this.hostname) {
                      this.stats.externalLinksFound++;
                    }
                  }
                } catch {}
                
                if (this.shouldFollowUrl(newUrl)) {
                  if (this.isAssetUrl(newUrl)) {
                    await this.downloadFile(newUrl);
                  } else {
                    this.queue.add(newUrl);
                  }
                }
              }
            }
          } catch (error) {
            this.log(`  ⚠ Error processing ${url}`);
          }
        });
        
        promises.push(promise);
      }
      
      await Promise.all(promises);
      promises.length = 0;
    }
    
    console.log('\n');
    this.printSummary();
  }

  printSummary() {
    const elapsed = (Date.now() - this.stats.startTime) / 1000;
    const minutes = Math.floor(elapsed / 60);
    const seconds = Math.floor(elapsed % 60);
    
    console.log('='.repeat(70));
    console.log('📊 CRAWL SUMMARY');
    console.log('='.repeat(70));
    console.log(`✅ HTML Pages:           ${this.stats.pagesDownloaded}`);
    console.log(`📥 Total Files:          ${this.stats.filesDownloaded}`);
    console.log(`💾 Data Saved:           ${(this.stats.bytesSaved / 1024 / 1024).toFixed(2)} MB`);
    console.log(`🌐 Domains Visited:      ${this.stats.domainsVisited}`);
    console.log(`🔗 External Links Found: ${this.stats.externalLinksFound}`);
    console.log(`❌ Errors:               ${this.stats.errors}`);
    console.log(`⏱️  Time:                 ${minutes}m ${seconds}s`);
    console.log(`📁 Saved to:             ${path.resolve(this.outputDir)}`);
    console.log('='.repeat(70));
    
    // File types breakdown
    if (Object.keys(this.fileTypes).length > 0) {
      console.log('\n📋 FILE TYPES SAVED:');
      const sorted = Object.entries(this.fileTypes)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10);
      
      for (const [ext, count] of sorted) {
        console.log(`  ${ext || '(no ext)'}: ${count} files`);
      }
      if (Object.keys(this.fileTypes).length > 10) {
        const remaining = Object.keys(this.fileTypes).length - 10;
        console.log(`  ... and ${remaining} more types`);
      }
    }
    
    // Errors
    if (this.errors.length > 0) {
      console.log('\n⚠️  FIRST 5 ERRORS:');
      this.errors.slice(0, 5).forEach(err => console.log(`  ${err}`));
      if (this.errors.length > 5) {
        console.log(`  ... and ${this.errors.length - 5} more`);
      }
    }
    
    console.log('='.repeat(70) + '\n');
  }
}

async function main() {
  try {
    console.clear();
    console.log('╔══════════════════════════════════════════════════════════════════╗');
    console.log('║                   🕷️  WEBSITE CRAWLER CLI  🕷️                     ║');
    console.log('║                                                                  ║');
    console.log('║  Download entire websites - all pages, assets, and more!         ║');
    console.log('╚══════════════════════════════════════════════════════════════════╝\n');
    
    // Get URL
    let url = process.argv[2];
    if (!url) {
      url = await question('🌍 Enter website URL: ');
    }
    
    if (!url.startsWith('http')) {
      url = 'https://' + url;
    }
    
    // Get options
    let outputDir = process.argv[3] || 'website_crawl';
    if (process.argv.length < 4) {
      const custom = await question(`\n📁 Output directory (default: ${outputDir}): `);
      if (custom.trim()) outputDir = custom.trim();
    }
    
    let followExternal = true;
    if (process.argv.length < 5) {
      const response = await question('\n🔗 Follow external links? (y/n, default: y): ');
      followExternal = response.toLowerCase() !== 'n';
    }
    
    let maxConcurrent = 5;
    if (process.argv.length < 6) {
      const response = await question('\n⚡ Max concurrent downloads (1-20, default: 5): ');
      const num = parseInt(response);
      if (num > 0 && num <= 20) maxConcurrent = num;
    }
    
    let maxPages = null;
    if (process.argv.length < 7) {
      const response = await question('\n📄 Max pages to crawl (leave empty for unlimited): ');
      const num = parseInt(response);
      if (num > 0) maxPages = num;
    }
    
    rl.close();
    
    // Start crawling
    const crawler = new AdvancedWebsiteCrawler(url, {
      outputDir,
      followExternal,
      maxConcurrent,
      maxPages,
      verbose: true
    });
    
    await crawler.crawl();
    
  } catch (error) {
    console.error('\n❌ Error:', error.message);
    process.exit(1);
  }
}

main();

File: crawler-files/crawl.py
#!/usr/bin/env python3
"""
Interactive Website Crawler - Python Version
Download any website completely - all pages and external links
"""

import os
import sys
import time
import json
import threading
from urllib.parse import urljoin, urlparse, unquote
from pathlib import Path
from collections import deque
from concurrent.futures import ThreadPoolExecutor, as_completed
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
from bs4 import BeautifulSoup
import mimetypes
from queue import Queue

class InteractiveWebsiteCrawler:
    def __init__(self, start_url, output_dir="website_crawl", follow_external=True, 
                 max_pages=None, max_workers=5, verbose=True):
        """
        Initialize crawler with support for external links
        """
        self.start_url = start_url.rstrip('/')
        self.output_dir = Path(output_dir)
        self.follow_external = follow_external
        self.max_pages = max_pages
        self.max_workers = max_workers
        self.verbose = verbose
        
        # Parse domain
        parsed = urlparse(start_url)
        self.domain = f"{parsed.scheme}://{parsed.netloc}"
        self.base_domain = parsed.netloc
        self.base_scheme = parsed.scheme
        
        # Tracking
        self.visited_urls = set()
        self.queue = deque([start_url])
        self.errors = []
        self.domains_found = {parsed.netloc}
        self.file_types = {}
        
        # Stats
        self.stats = {
            'pages_crawled': 0,
            'files_downloaded': 0,
            'bytes_saved': 0,
            'domains_visited': 1,
            'external_links': 0,
            'errors': 0,
            'start_time': time.time()
        }
        
        # Thread safety
        self.lock = threading.Lock()
        
        # Create output directory
        self.output_dir.mkdir(parents=True, exist_ok=True)
        
        # Setup requests session
        self.session = requests.Session()
        retry_strategy = Retry(
            total=3,
            backoff_factor=1,
            status_forcelist=[429, 500, 502, 503, 504],
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) WebsiteCrawler/2.0'
        })

    def log(self, msg):
        if self.verbose:
            print(msg)

    def is_valid_url(self, url):
        """Check if URL is valid"""
        if not url or url.startswith('javascript:') or url.startswith('mailto:') or url.startswith('data:'):
            return False
        
        try:
            parsed = urlparse(url)
            return parsed.scheme in ('http', 'https')
        except:
            return False

    def should_follow_url(self, url):
        """Check if we should follow this URL"""
        if not self.is_valid_url(url):
            return False
        
        if self.follow_external:
            return True
        
        # Stay on domain
        parsed = urlparse(url)
        return parsed.netloc == self.base_domain

    def get_local_path(self, url):
        """Convert URL to local file path"""
        parsed = urlparse(url)
        path = parsed.path.strip('/')
        
        # Handle root
        if not path:
            path = 'index.html'
        
        # Handle query strings
        if parsed.query:
            query_safe = parsed.query.replace('&', '_').replace('=', '_')[:30]
            base, ext = os.path.splitext(path)
            if ext:
                path = f"{base}_{query_safe}{ext}"
            else:
                path = f"{path}_{query_safe}.html"
        
        # Add .html if no extension
        if not os.path.splitext(path)[1]:
            path = f"{path}/index.html"
        
        full_path = self.output_dir / path
        return full_path

    def get_file_type(self, url):
        """Get file extension"""
        parsed = urlparse(url)
        path = parsed.path.lower()
        ext = os.path.splitext(path)[1] or 'unknown'
        return ext

    def is_asset_url(self, url):
        """Check if URL is an asset (not HTML)"""
        asset_extensions = {
            '.css', '.js', '.png', '.jpg', '.jpeg', '.gif', '.svg', '.webp',
            '.woff', '.woff2', '.ttf', '.otf', '.eot', '.mp4', '.webm',
            '.pdf', '.zip', '.ico', '.xml', '.json', '.mp3', '.tar', '.gz'
        }
        ext = self.get_file_type(url)
        return ext in asset_extensions

    def download_file(self, url):
        """Download a single file"""
        if url in self.visited_urls:
            return None
        
        with self.lock:
            if url in self.visited_urls:
                return None
            self.visited_urls.add(url)
        
        try:
            response = self.session.get(url, timeout=15, stream=True)
            response.raise_for_status()
            
            if response.status_code >= 400:
                return None
            
            local_path = self.get_local_path(url)
            local_path.parent.mkdir(parents=True, exist_ok=True)
            
            # Write file
            with open(local_path, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    if chunk:
                        f.write(chunk)
            
            file_type = self.get_file_type(url)
            with self.lock:
                self.file_types[file_type] = self.file_types.get(file_type, 0) + 1
                self.stats['files_downloaded'] += 1
                self.stats['bytes_saved'] += len(response.content)
            
            self.log(f"  ✓ {url[:80]}")
            return response
            
        except Exception as e:
            error_msg = f"✗ {url[:80]} ({str(e)[:40]})"
            self.log(f"  {error_msg}")
            with self.lock:
                self.errors.append(error_msg)
                self.stats['errors'] += 1
            return None

    def extract_urls_from_html(self, html_content, base_url):
        """Extract all URLs from HTML"""
        urls = set()
        
        try:
            soup = BeautifulSoup(html_content, 'html.parser')
            
            # Links
            for tag in soup.find_all(['a']):
                href = tag.get('href', '').strip()
                if href and not href.startswith('#'):
                    try:
                        absolute_url = urljoin(base_url, href).split('#')[0]
                        if self.should_follow_url(absolute_url):
                            urls.add(absolute_url)
                    except:
                        pass
            
            # Assets
            for tag in soup.find_all(['script', 'link', 'img', 'source', 'video', 'audio']):
                src = tag.get('src') or tag.get('href', '').strip()
                if src:
                    try:
                        absolute_url = urljoin(base_url, src)
                        if self.should_follow_url(absolute_url):
                            urls.add(absolute_url)
                    except:
                        pass
            
            # Srcset
            for tag in soup.find_all(['img']):
                srcset = tag.get('srcset', '').strip()
                if srcset:
                    for src_info in srcset.split(','):
                        src = src_info.strip().split()[0]
                        try:
                            absolute_url = urljoin(base_url, src)
                            if self.should_follow_url(absolute_url):
                                urls.add(absolute_url)
                        except:
                            pass
            
            # Data attributes
            for tag in soup.find_all(attrs={'data-src': True}):
                src = tag.get('data-src')
                try:
                    absolute_url = urljoin(base_url, src)
                    if self.should_follow_url(absolute_url):
                        urls.add(absolute_url)
                except:
                    pass
            
        except Exception as e:
            self.log(f"  ⚠ Error parsing HTML from {base_url}")
        
        return urls

    def crawl(self):
        """Main crawl loop with threading"""
        print('\n' + '='*70)
        print('🚀 WEBSITE CRAWLER STARTED')
        print('='*70)
        print(f'📍 Start URL:      {self.start_url}')
        print(f'📁 Output Dir:     {self.output_dir.absolute()}')
        print(f'🌍 Mode:           {"Follow ALL external links" if self.follow_external else "Stay on domain only"}')
        print(f'⚡ Concurrency:    {self.max_workers} simultaneous downloads')
        print('='*70 + '\n')
        
        batch_num = 0
        
        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
            while self.queue:
                if self.max_pages and self.stats['pages_crawled'] >= self.max_pages:
                    print(f'\n⏸️  Reached max pages limit ({self.max_pages})')
                    break
                
                batch_num += 1
                batch = []
                
                # Get batch
                for _ in range(min(self.max_workers, len(self.queue))):
                    if self.queue:
                        batch.append(self.queue.popleft())
                
                print(f'\n📦 Batch {batch_num} (Queue: {len(self.queue)}, Downloaded: {self.stats["files_downloaded"]})')
                
                futures = {}
                
                for url in batch:
                    future = executor.submit(self.download_file, url)
                    futures[future] = url
                
                # Process results
                for future in as_completed(futures):
                    url = futures[future]
                    response = future.result()
                    
                    if not response:
                        continue
                    
                    # Check if HTML
                    content_type = response.headers.get('content-type', '')
                    if 'text/html' not in content_type:
                        continue
                    
                    with self.lock:
                        self.stats['pages_crawled'] += 1
                    
                    # Extract URLs
                    try:
                        html = response.content.decode('utf-8', errors='ignore')
                        new_urls = self.extract_urls_from_html(html, url)
                        
                        for new_url in new_urls:
                            if new_url not in self.visited_urls:
                                # Track domains
                                try:
                                    parsed = urlparse(new_url)
                                    if parsed.netloc not in self.domains_found:
                                        with self.lock:
                                            self.domains_found.add(parsed.netloc)
                                            self.stats['domains_visited'] += 1
                                            if parsed.netloc != self.base_domain:
                                                self.stats['external_links'] += 1
                                except:
                                    pass
                                
                                if self.is_asset_url(new_url):
                                    executor.submit(self.download_file, new_url)
                                else:
                                    self.queue.append(new_url)
                    except Exception as e:
                        self.log(f'  ⚠ Error processing {url}')
        
        print('\n')
        self.print_summary()

    def print_summary(self):
        """Print summary statistics"""
        elapsed = time.time() - self.stats['start_time']
        minutes = int(elapsed // 60)
        seconds = int(elapsed % 60)
        
        print('='*70)
        print('📊 CRAWL SUMMARY')
        print('='*70)
        print(f"✅ HTML Pages:           {self.stats['pages_crawled']}")
        print(f"📥 Total Files:          {self.stats['files_downloaded']}")
        print(f"💾 Data Saved:           {self.stats['bytes_saved'] / 1024 / 1024:.2f} MB")
        print(f"🌐 Domains Visited:      {self.stats['domains_visited']}")
        print(f"🔗 External Links Found: {self.stats['external_links']}")
        print(f"❌ Errors:               {self.stats['errors']}")
        print(f"⏱️  Time:                 {minutes}m {seconds}s")
        print(f"📁 Saved to:             {self.output_dir.absolute()}")
        print('='*70)
        
        # File types
        if self.file_types:
            print('\n📋 FILE TYPES SAVED:')
            sorted_types = sorted(self.file_types.items(), key=lambda x: x[1], reverse=True)
            for file_type, count in sorted_types[:10]:
                print(f'  {file_type or "(no ext)"}: {count} files')
            if len(self.file_types) > 10:
                print(f'  ... and {len(self.file_types) - 10} more types')
        
        # Errors
        if self.errors:
            print('\n⚠️  FIRST 5 ERRORS:')
            for error in self.errors[:5]:
                print(f'  {error}')
            if len(self.errors) > 5:
                print(f'  ... and {len(self.errors) - 5} more')
        
        print('='*70 + '\n')

def get_input(prompt, default=""):
    """Get user input with optional default"""
    if default:
        prompt = f"{prompt} (default: {default}): "
    else:
        prompt = f"{prompt}: "
    
    response = input(prompt).strip()
    return response if response else default

def main():
    try:
        # Title
        print('\n╔══════════════════════════════════════════════════════════════════╗')
        print('║                   🕷️  WEBSITE CRAWLER CLI  🕷️                     ║')
        print('║                                                                  ║')
        print('║  Download entire websites - all pages, assets, and external links║')
        print('╚══════════════════════════════════════════════════════════════════╝')
        
        # Get URL
        url = get_input('\n🌍 Enter website URL', 'https://example.com')
        if not url.startswith('http'):
            url = 'https://' + url
        
        # Output dir
        output_dir = get_input('\n📁 Output directory', 'website_crawl')
        
        # Follow external
        follow_external_input = get_input('\n🔗 Follow external links? (y/n)', 'y').lower()
        follow_external = follow_external_input != 'n'
        
        # Max workers
        max_workers_input = get_input('⚡ Max concurrent downloads (1-20)', '5')
        try:
            max_workers = int(max_workers_input)
            max_workers = max(1, min(20, max_workers))
        except:
            max_workers = 5
        
        # Max pages
        max_pages_input = get_input('📄 Max pages (leave empty for unlimited)', '')
        try:
            max_pages = int(max_pages_input) if max_pages_input else None
        except:
            max_pages = None
        
        # Start crawling
        crawler = InteractiveWebsiteCrawler(
            url,
            output_dir=output_dir,
            follow_external=follow_external,
            max_pages=max_pages,
            max_workers=max_workers,
            verbose=True
        )
        
        crawler.crawl()
        
    except KeyboardInterrupt:
        print('\n\n⚠️  Crawl interrupted by user')
        sys.exit(0)
    except Exception as e:
        print(f'\n❌ Error: {e}')
        sys.exit(1)

if __name__ == "__main__":
    main()
